core/src/main/java/net/sourceforge/jnlp/util/LocaleUtils.java

```diff
@@ -0,0 +1,88 @@
+package net.sourceforge.jnlp.util;
+
+import java.util.Locale;
+import java.util.Objects;
+import net.adoptopenjdk.icedteaweb.xmlparser.ParseException;
+
+import static net.adoptopenjdk.icedteaweb.i18n.Translator.R;
+
+public class LocaleUtils {
+
+    public enum Match { LANG_COUNTRY_VARIANT, LANG_COUNTRY, LANG, GENERALIZED }
+
+    /**
+     * Returns a {@link Locale} from a locale string. Each locale is specified by a language identifier,
+     * possibly country identifier, and possibly a variant. The syntax is as follows:
+     * <pre>
+     * locale ::= language [ "_" country [ "_" variant ] ]
+     * </pre>
+     *
+     * @param localeString the locale string
+     * @return locale of document
+     */
+    public static Locale getLocale(final String localeString) throws ParseException {
+        if (Objects.isNull(localeString) || localeString.length() < 2) {
+            throw new ParseException(R("PIllegalLocale", localeString));
+        }
+
+        final String language = localeString.substring(0, 2);
+        final String country = (localeString.length() < 5) ? "" : localeString.substring(3, 5);
+        final String variant = (localeString.length() > 7) ? localeString.substring(6) : "";
+
+        // null is not allowed n locale but "" is
+        return new Locale(language, country, variant);
+    }
+
+    /**
+     * Returns whether a locale is matched by one of more other
+     * locales. Only the non-empty language, country, and variant
+     * codes are compared; for example, a requested locale of
+     * Locale("","","") would always return true.
+     *
+     * @param requested the requested locale
+     * @param available the available locales
+     * @param matchLevel the depth with which to match locales.
+     * @return {@code true} if {@code requested} matches any of {@code available}, or if
+     * {@code available} is empty or {@code null}.
+     * @see Locale
+     * @see Match
+     */
+    public static boolean localeMatches(Locale requested, Locale[] available, Match matchLevel) {
+
+        if (matchLevel == Match.GENERALIZED)
```

Comment:
can be moved to the switch statement below as a case for GENERALIZED

Reply:
does not work, as we have a unit test that explicitly defines GENERALIZED and null for available. Which leads to NPE in for loop when moving GENERALIZED to switch. As because of null for available and not empty array in case of no locales this needs bigger refactoring for all callers to ensure they handle correctly, e.g. hand in empty array and localMatches then asserts available to be non null. So keep it for now.

But I added some code as your comment pointed me to another potential problem: available=null and any other match level would end in NPE.

Reply:
done